index.d.ts

```diff
@@ -0,0 +1,295 @@
+       * @param filterer Async function. Gets passed (value, index, array),
+       * returns true to keep the value in the resulting filtered Array.
+       * @returns Array resulting filtered values.
+       */
+      filter<T>(
+        array: ArrayLike<T>,
+        filterer: (value: T, index: number, array: T[]) => PromiseLike<any>,
+      ): Promise<T[]>;
+      /**
+       * Invoke an async function on each item in the given Array in parallel,
+       * returning the first element predicate returns truthy for.
+       * > This is an asynchronous, parallelized version of
+       * Array.prototype.find().
+       * @param array The Array to find.
+       * @param predicate Async function. Gets passed (value, index, array), returns true
+       * to be the find result.
+       * @returns Any resulting find value.
+       */
+      find<T>(
+        array: ArrayLike<T>,
+        predicate: (value: T, index: number, array: T[]) => PromiseLike<boolean>,
+      ): Promise<T | undefined>;
+  
+      /**
+       * Checks if predicate returns truthy for all elements of collection in
+       * parallel.
+       * > This is an asynchronous, parallelized version of
+       * Array.prototype.every().
+       * @param array The Array to iterate over.
+       * @param predicate Async function. Gets passed (value, index, array), The
+       * function invoked per iteration.
+       * @returns Returns true if all element passes the predicate check, else
+       * false.
+       */
+      every<T>(
+        array: ArrayLike<T>,
+        predicate: (value: T, index: number, array: T[]) => PromiseLike<boolean>,
+      ): boolean;
+  
+      /**
+       * Checks if predicate returns truthy for any element of collection in
+       * parallel.
+       * @param array The Array to iterate over.
+       * @param predicate Async function. Gets passed (value, index, array), The
+       * function invoked per iteration.
+       * @returns Returns true if any element passes the predicate check, else
+       * false.
+       */
+      some<T>(
+        array: ArrayLike<T>,
+        predicate: (value: T, index: number, array: T[]) => PromiseLike<boolean>,
+      ): boolean;
+  
+      /**
+       * Invoke all async functions in an Object in parallel, returning the
+       * result.
+       * @param list Object with values that are async functions to invoke.
+       * @returns Same structure as list input, but with values now resolved.
+       */
+      parallel<T extends Record<string, AsyncFn<any>>>(
+        list: T,
+      ): Promise<{[K in keyof T]: AsyncFnReturnType<T[K]>}>;
+  
+      /**
+       * Invoke all async functions in an Array in parallel, returning the
+       * result.
+       * @param list Array with values that are async functions to invoke.
+       * @returns Same structure as list input, but with values now resolved.
+       */
+      parallel<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(
+        list: [
+          AsyncFn<T1>,
```

Comment:
Is there a way to avoid having to cap the list here?

Reply:
In fact, yes, but only if we use rest parameters instead of a single `list` parameter.

With the introduction of [rest parameters with tuple types in TypeScript 3.0](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#rest-parameters-with-tuple-types) and [mapped types on tuples in TypeScript 3.1](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays), we can simplify the type definition for the `parallel` method to this:

```ts
series<T extends any[]>(
  ...list: T
): Promise<{[K in keyof T]: AsyncFnReturnType<T[K]>}>;
```
But then, it's probably not an option to change the function signature.

---

However, with the introduction of [optional elements in tuple types in TypeScript 3.0](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#rest-parameters-with-tuple-types#optional-elements-in-tuple-types), we can get rid of overloads and reduce the method's type definition to:

```ts
series<
  T extends [any, any?, any?, any?, any?, any?, any?, any?, any?, any?]
>(
  list: T,
): Promise<{[K in keyof T]: AsyncFnReturnType<T[K]>}>;
```

Then, we can increase the capping to a higher number of elements (by adding extra `any?` elements to the generic tuple type) without adding a ton of overloads.

Reply:
@developit @yenbekbay any chance to get a new version with the updated typings? This is really a blocker for those who use Typescript :(

Reply:
TypeScript is version 3.7 now. Maybe we can do something.

cc @developit